source/language/types.rst

```diff
@@ -373,3 +375,174 @@ variables whose values may only be known at run time.
+| **bit**      | :gbg:`Yes` | :gbg:`Yes` | :gbg:`Yes` | :rbg:`No`   | :gbg:`Yes` | :center:`-`| :rbg:`No`  | :rbg:`No`  |
++--------------+------------+------------+------------+-------------+------------+------------+------------+------------+
+| **duration** | :rbg:`No`  | :rbg:`No`  | :rbg:`No`  | :rbg:`No*`  | :rbg:`No`  | :rbg:`No`  | :center:`-`| :rbg:`No`  |
++--------------+------------+------------+------------+-------------+------------+------------+------------+------------+
+| **qubit**    | :rbg:`No`  | :rbg:`No`  | :rbg:`No`  | :rbg:`No`   | :rbg:`No`  | :rbg:`No`  | :rbg:`No`  | :center:`-`|
++--------------+------------+------------+------------+-------------+------------+------------+------------+------------+
+
+\*Note: ``duration`` values can be converted to ``float`` using the division operator. See :ref:`divideDuration`
+
+Casting from bool
+~~~~~~~~~~~~~~~~~
+
+``bool`` values cast from ``false`` to ``0.0`` and from ``true`` to ``1.0`` or 
+an equivalent representation. ``bool`` values can only be cast to ``bit[1]`` 
+(a single bit), so explicit index syntax must be given if the target ``bit``
+has more than 1 bit of precision.
+
+Casting from int/uint
+~~~~~~~~~~~~~~~~~~~~~
+
+``int[n]`` and ``uint[n]`` values cast to the standard types mimicking C99
+behavior. Casting to ``bool`` values follows the convention ``val != 0``.
+As noted above, if the value is too large to be represented in the
+target type the result is implementation-specific. Casting to ``bit[m]`` 
+is only allowed when ``m==n``. If the target 
+``bit`` has more or less precision, then explicit slicing syntax must be given.
+As noted, the conversion is done assuming a little-endian 2's complement 
+representation.
+
+Casting from float
+~~~~~~~~~~~~~~~~~~
+
+``float[n]`` values cast to the standard types mimicking C99 behavior (*e.g.* 
+discarding the fractional part for integer-type targets). As noted above, 
+if the value is too large to be represented in the 
+target type the result is implementation-specific. 
+Casting a ``float`` value to an ``angle[m]`` is accomplished by first 
+performing a modulo 2π operation on the float value. The resulting value 
+is then converted to the nearest ``angle[m]`` possible. In the event of a 
+tie between two possible nearest values the result is the one with an even 
+least significant bit (*i.e.* round to nearest, ties to even).
+
+Casting from angle
+~~~~~~~~~~~~~~~~~~
+
+``angle[n]`` values cast to ``bool`` using the convention ``val != 0.0``. 
+Casting to ``bit[m]`` values is only allowed when 
+``n==m``, otherwise explicit slicing synatx must be provided.
+Casting to ``float[m]`` is accomplished by first converting the angle value 
+to the nearest float, with ties going to the value with the even least 
+significant bit, and then multiplying the resulting value by 2π as a 
+``float[m]`` value. Note: the ring-like behavior of the angle type can
```

Comment:
Can you suggest some syntax that would enable specifying what type of cast the user wants? Should we use a pragma? Can each cast be different, or is it a global setting?

Reply:
I would suggest syntax along the lines of ``angleToSignedFloat(angle a)`` and ``angleToUnsignedFloat(angle a)``, or possibly ``angleToFloatBelow(angle a, float u)`` to obtain a result ``x`` satisfying ``u - 2*π <= x < u``. (These names are perhaps not the best possible choices; I just mean to convey the meaning.) I have become wary of the idea of allowing the syntax of ``(float)a``, because it suggests a single standard approach to conversion, and I no longer believe that this is a good idea. Just conceivably one might want to allow a pragma to allow the programmer to fix the meaning of ``(float)a`` to some function as above using a pragma, but I am not even sure that this is a good idea.

Reply:
If syntax such as ``(signed)a`` or ``(unsigned)a`` were permissable, this might represent a good choice; though it suggests the existence of types which don't exist. ``signedf(a)`` or ``unsignedf(a)`` might be closer to a good choice of syntax.

Reply:
Interesting idea, @ndebeaudrap

Reply:
@blakejohnson We don't _have_ to define what ``float(a)`` means for ``angle a``, or at least not any more than we have to define what ``angle(b)`` means for ``bool b``. We can decide that it does not make sense, after all, to allow this as a typecast _as such_. What I advocate is to disallow the typecast, but instead provide functions (which may be evaluated in some instances at compile time) which take an ``angle`` as input and produce a ``float``, according to various specifications.

Having said that: your proposal allowing ``float(...)`` to take additional arguments is interesting. I would suggest that we should be extremely careful of allowing these arguments to be optional, in the case that the first argument is to be of type ``angle``. If we were to take this approach, I would advocate syntax along the lines of ``float(angle a, float ivlCenter)``, where the second argument is mandatory and indicates the center of a half-open interval of width 2π. For instance, ``float(a, 0)`` might produce a float in the range [-π,π), while ``float(a,π)`` might produce a float in the range [0,2π).

Reply:
The discussion at the TG meeting resolved in us deciding to remove the possibility of casting directly from angle to float from the spec for now. We will try to address it more fully in a future PR.